### DISCUSSION TITLE
Solving a least squares rather than a quadratic program?

### DISCUSSION THREAD

**Discussion by stephane-caron** [qpsolvers/qpsolvers#223] [Q&A] [answered]
Is it possible to solve a least squares rather than a quadratic program?

**Comment by stephane-caron** [comment, marked as answer]
Yes, there is also a [`solve_ls`](https://qpsolvers.github.io/qpsolvers/least-squares.html#qpsolvers.solve_ls) function.

See also the post [Conversion from least squares to quadratic programming](https://scaron.info/blog/conversion-from-least-squares-to-quadratic-programming.html) for details on how this is done, and inspiration if you need to convert a similar problem.

**Comment by ogencoglu** [comment]
How about ridge regression or LASSO, i.e., having an additional penalty term in the objective function regarding the magnitude of the weights?

**Reply by stephane-caron** [reply]
Ridge regression and Lasso can both be formulated as quadratic programming. The library does not have `solve_ridge` or `solve_lasso` functions to do the conversion automatically (PRs welcome), but you can call [`solve_qp`](https://qpsolvers.github.io/qpsolvers/quadratic-programming.html#qpsolvers.solve_qp) directly.

**Reply by ogencoglu** [reply]
Thanks for the swift reply! Is there really a way to formulate arbitrary penalty with `solve_qp` though? Let's say `L_p` penalty where p is 1.4 or whatever (L2 is ridge, L1 is Lasso).

**Reply by stephane-caron** [reply]
You're welcome. Here is a more detailed answer:

- **Ridge regression:** yes, $\ell^2$ norm goes to the cost matrix $P$.
- **Lasso:** yes, $\ell^1$ norm goes to the cost vector $q$ and inequality constraints. Essentially one introduces a new variable $z = |x|$ via inequality constraints: $x \leq z$ and $-x \leq z$, then minimize $-1^T z$ in the cost (where $1$ is a ones vector) so that all inequalities are tight. Then $z = |x|$ at the optimum and $\|x\|_1 \leq t$ with a constant parameter $t$ is equivalent to adding the inequality constraint $z \leq t$.
- **$\ell^p$ norm:** I'm not an expert on this and have not thought this question through, so my answer won't help much :wink: My hunch is "not always", but I let you research this question properly.

**Reply by stephane-caron** [reply]
@ogencoglu I wrote the following note on [Lasso regularization in quadratic programming](https://scaron.info/blog/lasso-regularization-in-quadratic-programming.html). Could you read it and let me know whether it is correct and clear, or if you have other feedback?

**Reply by ogencoglu** [reply]
Thanks a lot for the quick write-up. It seems indeed correct. I also agree that fractional or non-integer L-norms would not work in QP context.

**Reply by ogencoglu** [reply]
A side note: typically machine learning people use lambda to denote `t` in your equation.

**Reply by stephane-caron** [reply]
Thank you for your feedback :+1: I see in research articles (for instance [Complexity Analysis of the Lasso Regularization Path](https://arxiv.org/pdf/1205.0079.pdf)) that $\lambda$ is used in the objective function rather than in the inequality constraints, and the resulting formulation is also called "lasso". That is not exactly the formulation from the original article by [Tibshirani (1996)](https://doi.org/10.1111/j.2517-6161.1996.tb02080.x) which is the one followed on the Wikipedia article on [Lasso (statistics)](https://en.wikipedia.org/wiki/Lasso_(statistics)). But the $\lambda$ is also mentioned in the Wikipedia article in [Making λ easier to interpret with an accuracy-simplicity tradeoff](https://en.wikipedia.org/wiki/Lasso_(statistics)#Making_%CE%BB_easier_to_interpret_with_an_accuracy-simplicity_tradeoff).

TL;DR: the note now mentions both $t$ and $\lambda$ :wink:

**Comment by ogencoglu** [comment]
Another related question:

In order to optimize the intercept term (i.e., `b` in `ax + b`) as well in the regression, does the following makes sense:

- appending a column of 1s to the end of `R` and
- appending a column of zeros to the end of `A`

Do you see any issues with this approach or would you achieve the same in another way?

**Reply by stephane-caron** [reply]
Hi @ogencoglu, I will take a look at your question, but could you open a separate discussion for each new question?

This will make it easier for future travelers to find question and answers.

**Reply by ogencoglu** [reply]
Will do, thanks.
